client/my-sites/plans/index.js

```diff
@@ -10,38 +10,89 @@ import { features, plans, redirectToCheckout, redirectToPlans } from './controll
 import { currentPlan } from './current-plan/controller';
 import { makeLayout, render as clientRender } from 'controller';
 import { navigation, siteSelection, sites } from 'my-sites/controller';
+import { trackNavigationStart } from 'lib/performance-tracking';
 
 export default function () {
-	page( '/plans', siteSelection, sites, makeLayout, clientRender );
-	page( '/plans/compare', siteSelection, navigation, redirectToPlans, makeLayout, clientRender );
+	page( '/plans', trackNavigationStart( 'plans' ), siteSelection, sites, makeLayout, clientRender );
+	page(
+		'/plans/compare',
+		trackNavigationStart( 'plans' ),
```

Comment:
All these routes share the same label `plans`. They will appear as the same page in Speedcurve.

We can make this as granular as we want, but AFAIK Speedcurve doesn't provide a way to aggregate labels later.

Reply:
I wonder if we shouldn't just use the same id as the route passed to page?

Rather than duplicating the route names everywhere, could we create a wrapper/"decorator" for page that automatically does this.. something like:

```js
const trackNavigation = ( page ) => ( routeName, ...rest ) =>
    page( routeName, trackNavigationStart( routeName ), ...rest );
```

While we're at it... I wonder if there's a way to also automatically incorporate `trackNavigationStop`, perhaps by utilizing the last callback parameter for `ReactDom.render` 🤔

A generic way to do it could be to add an `onRendered` array to the `context` in the `trackNavigationStart` middleware which could be consumed by `clientRender`...
Along with that we could add a `onRendered` array to `context` with a list of callbacks to fire in `ReactDom.render`.

```js
const trackNavigationStart = ( context, next ) => {
  // ...

  context.onRendered = context.onRendered || [];
  context.onRendered.push( () => trackNavigationEnd() );
};

function render( context ) {
	ReactDom.render(
		context.layout,
		document.getElementById( 'wpcom' ),
		() => {
			if ( Array.isArray( context.onRendered ) ) {
				context.onRendered.forEach( fn => fn() );
			}
		},
	);
}
```

I think that would eliminate having a whole set of steps you'd need to complete to get correct tracking for a page... you could just decorate `page` with `trackNavigation` and then call it a day! Makes it easier to remove it too, less to clean up per route.

Reply:
> I wonder if we shouldn't just use the same id as the route passed to page?

I think that would be too granular, given that Speedcurve doesn't seem to have a way to aggregate them later (eg: 'show me all /plans/*' pages)

> Rather than duplicating the route names everywhere, could we create a wrapper/"decorator" for page that automatically does this

Good idea, I'll try it

> While we're at it... I wonder if there's a way to also automatically incorporate `trackNavigationStop`

I don't think so. The router doesn't know when a full page is ready. For example, if everything the page does is render an `<AsyncLoad>` component, `trackNavigationStop` should be called when the "internal" component is loaded and rendered, not when the `<AsyncLoad>` is rendered.

Another shortcoming of the current approach is that it doesn't provide a way to express "this page is ready when components X and Y are rendered" (which would be useful in the 'stats' page).

Reply:
> For example, if everything the page does is render an <AsyncLoad> component, trackNavigationStop should be called when the "internal" component is loaded and rendered, not when the <AsyncLoad> is rendered

But `ReactDom.render` would presumably know about it for the callback right? But maybe not for a situation you're describing like `AsyncLoad` 🤔

But I see what you're saying, kind of any solution requires answering a question like "what is the lowest, furthest down page component that will be loaded when the initial page load is ready?" If we're not putting the "stop" at a deep enough level, then we're still not getting an accurate TTR, as in falling short of it... And two `AsyncLoad`s next to each other aren't necessarily going to resolve around the same time either. 😰